packages/network-controller/src/provider-api-tests/shared-tests.ts

```diff
@@ -0,0 +1,2673 @@
+            response: {
+              result: 'the result',
+              httpStatus: 200,
+            },
+          });
+          const result = await withNetworkClient(
+            { providerType },
+            async ({ makeRpcCall, clock }) => {
+              return await waitForPromiseToBeFulfilledAfterRunningAllTimers(
+                makeRpcCall(request),
+                clock,
+              );
+            },
+          );
+
+          expect(result).toStrictEqual('the result');
+        });
+      });
+
+      // Both the Infura middleware and custom RPC middleware detect a 503 or 504
+      // response and retry the request to the RPC endpoint automatically but
+      // differ in what sort of response is returned when the number of retries is
+      // exhausted.
+      if (providerType === 'infura') {
+        it(`causes a request to fail with a custom error if the request to the RPC endpoint returns a ${httpStatus} response 5 times in a row`, async () => {
+          await withMockedCommunications({ providerType }, async (comms) => {
+            const request = {
+              method,
+              params: buildMockParams({ blockParam, blockParamIndex }),
+            };
+
+            // The first time a block-cacheable request is made, the
+            // block-cache middleware will request the latest block number
+            // through the block tracker to determine the cache key. Later,
+            // the block-ref middleware will request the latest block number
+            // again to resolve the value of "latest", but the block number is
+            // cached once made, so we only need to mock the request once.
+            comms.mockNextBlockTrackerRequest({ blockNumber: '0x100' });
+            // The block-ref middleware will make the request as specified
+            // except that the block param is replaced with the latest block
+            // number.
+            comms.mockRpcCall({
+              request: buildRequestWithReplacedBlockParam(
+                request,
+                blockParamIndex,
+                blockParam === undefined ? null : 'latest',
+              ),
+              response: {
+                error: 'Some error',
+                httpStatus,
+              },
+              times: 5,
+            });
+            const promiseForResult = withNetworkClient(
+              { providerType },
+              async ({ makeRpcCall, clock }) => {
+                return await waitForPromiseToBeFulfilledAfterRunningAllTimers(
+                  makeRpcCall(request),
+                  clock,
+                );
+              },
+            );
+            await expect(promiseForResult).rejects.toThrow(
+              buildInfuraClientRetriesExhaustedErrorMessage('Gateway timeout'),
+            );
+          });
+        });
+      } else {
+        it(`produces an empty response if the request to the RPC endpoint returns a ${httpStatus} response 5 times in a row`, async () => {
```

Comment:
Nit: This test name isn't totally accurate, as it's no longer quite an empty response. Perhaps something like this would be more clear?

``` suggestion
        it(`produces a response without a result if the request to the RPC endpoint returns a ${httpStatus} response 5 times in a row`, async () => {
```

Same for the other instances.